### DISCUSSION TITLE
ATEM Network Proxy Server

### DISCUSSION THREAD

**Discussion by hrueger** [josephdadams/TallyArbiter#346] [Ideas]
Hi all,
this is definitely not the right place to discuss this feature as it is only indirectly related to TallyArbiter, but I'd like to have some thoughts of you about the following idea:

I really enjoy watching @aaronpk's weekly livestreams about video production and [here](https://youtu.be/sYmBjxDn4F0?t=4933) he says that the problem with using TallyArbiter for him is that he would run into the ATEM's concurrent connection limit of five devices. 

I have personally run into this, too (1 PC with TA, the ATEM Software Control and my own control software + a second PC with the ATEM Software Control and my control software as well --> a third PC could not connect anymore)

So I wondered if we can create a Proxy Server which could relay the commands sent to and received from the ATEM to all connected devices. Then, you would connect the proxy server to the ATEM and other devices to that proxy. Has anyone here thoughts on that?

It's probably not enough to just combine the TCP streams as we need to send the return values of commands back to the correct sender...

Maybe @julusian, who maintains the `atem-connection` library, has some thoughts if that is even possible?

**Comment by Julusian** [comment]
This is very possible to do, and I did start playing around with it last year https://github.com/LibAtem/AtemProxy
From memory, that application does work, but hasn't been tested enough for me to be confident in recommending its use.

In reality, the application is not that complicated. The protocol is pretty stateless, and everything generally should go to everyone. The exception being the locks and media transfer bits, and audio levels as they are opt-in and can be a lot of traffic

**Reply by hrueger** [reply]
Wow, that's cool. Unfortunately, I'm not very familiar with C#, so I can only help with testing.

**Reply by hrueger** [reply]
Would you mind providing binaries for testing?

**Reply by mattv8** [reply]
@hrueger, here's a binary for testing. It's a console app, so in Windows environment you'll need to open Command Prompt and run the following command: `/path/to/AtemProxy.exe 10.1.1.32` (replacing "10.1.1.32" with the IP address of your ATEM). There won't be much console output as unfortunately I wasn't able to get log4net to be fully verbose. Might need @Julusian's help with that.
[AtemProxy.zip](https://github.com/josephdadams/TallyArbiter/files/7858711/AtemProxy.zip)
Example output (with and without verbosity)
![Untitled](https://user-images.githubusercontent.com/9312603/149242057-85f489a3-1d8f-4ad5-9188-12cef3d4da5b.png)

I don't have an ATEM on hand to test, unfortunately.

**Comment by josephdadams** [comment]
This is a great idea. Skaarhoj has a product for this as well. We only have one ATEM here and we hardly use it. I could help test but probably not put a lot of time into development for it. @mattv8 is our Tally Arbiter C# man, let's rope him into this. :)

**Reply by mattv8** [reply]
Thanks for looping me in. I'll take a look at @Julusian's C# project and see if I can contribute in some way. At the very least I could compile some .exe's for @hrueger to test if Julusian doesn't get to it before me :)

**Reply by JTF4** [reply]
I've done some work in C#. Mostly GUI stuff in .net, but I can help out a little if needed.

I would also like to look into a node version for cross platform. With C# it can really be a pain to run on windows and mac.

**Reply by hrueger** [reply]
Raspberry Pi would also be cool. I don't know how easy / hard it is to compile C# for that platform.

**Reply by Julusian** [reply]
I have thrown some binaries up at https://github.com/LibAtem/AtemProxy/releases/tag/v0.1.0
I havent tested that any of them run, but hopefully they will.

@JTF4 I started the c# atem stuff soon after dotnetcore was released, and I have made sure to keep it completely cross platform.
The problem with doing a node version is that the application needs to understand every command that can pass through it, and needs to be able to extract an 'identifier' for each, so that it can send the correct state to clients upon them reconnecting without having to forward many hours worth of commands.

**Reply by hrueger** [reply]
Thanks @Julusian, I'll test them as soon as I can.
